src/rotation/log_new.zig

```diff
@@ -0,0 +1,340 @@
+			if(!connectedNegY) dir = .posYDir;
+			if(!connectedPosX) dir = .negXDir;
+
+			return .{.intersection = dir};
+		},
+		4 => {
+			return .cross;
+		},
+		else => undefined,
+	};
+}
+
+pub fn createBlockModel(_: Block, _: *u16, _: ZonElement) ModelIndex {
+	if(modelIndex) |idx| return idx;
+
+	for(0..64) |i| {
+		var quads = main.List(main.models.QuadInfo).init(main.stackAllocator);
+		defer quads.deinit();
+
+		const data = LogData.init(@intCast(i));
+
+		for(Neighbor.iterable) |neighbor| {
+			const pattern = getPattern(data, neighbor);
+
+			addQuads(pattern, neighbor, &quads);
+		}
+
+		const index = main.models.Model.init(quads.items);
+		if(i == 0) {
+			modelIndex = index;
+		}
+	}
+
+	return modelIndex.?;
+}
+
+pub fn model(block: Block) ModelIndex {
+	return .{.index = blocks.meshes.modelIndexStart(block).index + (block.data & 63)};
+}
+
+pub fn rotateZ(data: u16, angle: Degrees) u16 {
+	@setEvalBranchQuota(65_536);
+
+	comptime var rotationTable: [4][16]u8 = undefined;
+	comptime for(0..16) |i| {
+		rotationTable[0][i] = @intCast(i << 2);
+	};
+	comptime for(1..4) |a| {
+		for(0..16) |i| {
+			const old: LogData = .init(rotationTable[a - 1][i]);
+			var new: LogData = .init(0);
+
+			new.setConnection(Neighbor.dirPosX.rotateZ(), old.isConnected(Neighbor.dirPosX));
+			new.setConnection(Neighbor.dirNegX.rotateZ(), old.isConnected(Neighbor.dirNegX));
+			new.setConnection(Neighbor.dirPosY.rotateZ(), old.isConnected(Neighbor.dirPosY));
+			new.setConnection(Neighbor.dirNegY.rotateZ(), old.isConnected(Neighbor.dirNegY));
+
+			rotationTable[a][i] = new.enabledConnections;
+		}
+	};
+	if(data >= 0b111111) return 0;
+	const rotationIndex = (data & 0b111100) >> 2;
+	const upDownFlags = data & 0b000011;
+	return rotationTable[@intFromEnum(angle)][rotationIndex] | upDownFlags;
+}
+
+pub fn generateData(
+	_: *main.game.World,
+	pos: Vec3i,
+	_: Vec3f,
+	_: Vec3f,
+	_: Vec3i,
+	neighbor: ?Neighbor,
+	currentBlock: *Block,
+	neighborBlock: Block,
+	blockPlacing: bool,
+) bool {
+	const canConnectToNeighbor = currentBlock.mode() == neighborBlock.mode() and currentBlock.modeData() == neighborBlock.modeData();
+
+	if(blockPlacing or canConnectToNeighbor or !neighborBlock.replacable()) {
+		const neighborModel = blocks.meshes.model(neighborBlock).model();
+
+		var currentData = LogData.init(currentBlock.data);
+		// Log block upon placement should extend towards a block it was placed
+		// on if the block is solid or also uses log model.
+		const targetVal = ((!neighborBlock.replacable() and (!neighborBlock.viewThrough() or canConnectToNeighbor)) and (canConnectToNeighbor or neighborModel.isNeighborOccluded[neighbor.?.reverse().toInt()]));
+		currentData.setConnection(neighbor.?, targetVal);
+
+		for(Neighbor.iterable) |side| {
+			if(side == neighbor.?) {
+				continue;
+			}
+
+			const sidePos = pos + side.relPos();
+			const sideBlock = main.renderer.mesh_storage.getBlock(sidePos[0], sidePos[1], sidePos[2]) orelse continue;
```

Comment:
This will fail on headless servers, you cannot query blocks here.

Reply:
Also I don't quite understand why this code is doing. It wasn't there for branches.

Reply:
Ah, makes sense. It does seem a bit odd but I guess it's probably the best solution.
Also I discovered that the function is only called on the client, so it's fine.